### PR TITLE
Improve username format validation

### PR DESCRIPTION
**Pull request by Thisara-Welmilla**
Improve username format validation check at self registration

### CODE REVIEW

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java [hunk 1]
```diff
@@ -1471,6 +1481,44 @@ public boolean isMatchUserNameRegex(String tenantDomain, String username) throws
...
+        int tenantId = realmConfig.getTenantId();
+        String tenantDomain = IdentityTenantUtil.getTenantDomain(tenantId);
+        Map<String, Validator> validators = IdentityRecoveryServiceDataHolder.getInstance()
+                .getInputValidationMgtService().getValidators(tenantDomain);
+
+        List<ValidationConfiguration> configurations;
+        try {
+            configurations = IdentityRecoveryServiceDataHolder.getInstance().getInputValidationMgtService()
+                    .getInputValidationConfiguration(tenantDomain);
+            String field = USERNAME;
```
[the remark this answers is not in the document]
Comment (Thisara-Welmilla) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java:1492]: Already using `org.wso2.carbon.identity.input.validation.mgt.utils.Constants.Configs.USERNAME` constant

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java [hunk 2]
```diff
@@ -1497,7 +1545,36 @@ private boolean checkUserNameValid(String userName, RealmConfiguration realmConf
         }
 
         return StringUtils.isEmpty(regularExpression) || isFormatCorrect(regularExpression, userName);
+    }
 
+    /** This method is to validate username format against the username validation configured with Input Validation.
```
[the remark this answers is not in the document]
Comment (Thisara-Welmilla) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java:1549]: Addressed

File: components/org.wso2.carbon.identity.password.policy/src/main/java/org/wso2/carbon/identity/password/policy/internal/IdentityPasswordPolicyServiceComponent.java [hunk 3]
```diff
@@ -1,18 +1,21 @@
 /*
- * Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ * Copyright (c) 2023, WSO2 LLC. (https://www.wso2.com) All Rights Reserved.
```
Comment (kayathiri4) [inline components/org.wso2.carbon.identity.password.policy/src/main/java/org/wso2/carbon/identity/password/policy/internal/IdentityPasswordPolicyServiceComponent.java:2]: Is it intentional to update the license header to the latest year?

File: components/org.wso2.carbon.identity.password.policy/src/main/java/org/wso2/carbon/identity/password/policy/internal/IdentityPasswordPolicyServiceDataHolder.java [hunk 4]
```diff
@@ -1,17 +1,19 @@
 /*
- * Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ * Copyright (c) 2023, WSO2 LLC. (https://www.wso2.com) All Rights Reserved.
```
Comment (kayathiri4) [inline components/org.wso2.carbon.identity.password.policy/src/main/java/org/wso2/carbon/identity/password/policy/internal/IdentityPasswordPolicyServiceDataHolder.java:2]: Check the license header

Comment (kayathiri4) [reply components/org.wso2.carbon.identity.password.policy/src/main/java/org/wso2/carbon/identity/password/policy/internal/IdentityPasswordPolicyServiceDataHolder.java:2]: Check other places as well

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java [hunk 5]
```diff
@@ -1471,6 +1480,44 @@ public boolean isMatchUserNameRegex(String tenantDomain, String username) throws
...
+                return validateAgainstRegex(userName, realmConfig);
+            }
+        } catch(InputValidationMgtException e){
+            return ERROR_GETTING_EXISTING_CONFIGURATIONS.getCode().equals(e.getErrorCode());
+        }
+    }
+
+    /** This method is to validate username format against the regex from the userstore.
+     *
+     * @param userName      Username that needs to be validated
```
Comment (kayathiri4) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java:1515]:
```suggestion
     * @param userName      Username that needs to be validated.
```

Comment (kayathiri4) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java:1515]: Can you please check other places as well.

File: components/org.wso2.carbon.identity.password.policy/src/main/java/org/wso2/carbon/identity/password/policy/internal/IdentityPasswordPolicyServiceComponent.java (hunk 3 above)
Comment (Thisara-Welmilla) [reply components/org.wso2.carbon.identity.password.policy/src/main/java/org/wso2/carbon/identity/password/policy/internal/IdentityPasswordPolicyServiceComponent.java]: Addressed

File: components/org.wso2.carbon.identity.password.policy/src/main/java/org/wso2/carbon/identity/password/policy/internal/IdentityPasswordPolicyServiceDataHolder.java (hunk 4 above)
Comment (Thisara-Welmilla) [reply components/org.wso2.carbon.identity.password.policy/src/main/java/org/wso2/carbon/identity/password/policy/internal/IdentityPasswordPolicyServiceDataHolder.java]: Addressed

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java (hunk 5 above)
Comment (Thisara-Welmilla) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java]: Addressed

**Review by kayathiri4 — APPROVED** [review]

**Pull request merged**